### PR TITLE
Upgrade PMD to 7.0.0

### PR DESCRIPTION
**Pull request by holgerfriedrich**
Add support for Java 21.

https://docs.pmd-code.org/pmd-doc-7.0.0-rc4/pmd_release_notes.html

---

Update to PMD7 gives me some headache, as AST has changed.
See migration guide, https://docs.pmd-code.org/pmd-doc-7.0.0-rc4/pmd_release_notes.html#migration-guide-for-pmd-7

Still struggling to get custom test to work properly (UseSLF4JLoggerRule.java).

@wborn Could use some help, just in case you know the details of AST parsing.....

### CODE REVIEW

**Comment by wborn** [comment]
I had a look and after setting a break point I could check the objects and see where all the required data is stored nowadays.
Then I could get it working again with the changes in https://github.com/holgerfriedrich/static-code-analysis/pull/1.

**Comment by holgerfriedrich** [comment]
https://github.com/openhab/static-code-analysis/pull/463#issuecomment-1975116887

**Comment by holgerfriedrich** [comment]
benchmark looks good, removed wip label

**Comment by holgerfriedrich** [comment]
Hello @openhab/sat-maintainers , kindly asking to review/merge this PR. We need this as one of the last pieces towards a Java21 build. Thanks!

See the comparison of 0.15 and 0.16-SNAPSHOT: https://github.com/openhab/static-code-analysis/pull/463#issuecomment-1975116887 

See comments above:

>    btw: Once we get https://github.com/openhab/static-code-analysis/pull/462 and https://github.com/openhab/static-code-analysis/pull/463, we should release.
    The benefit it support for Java 21 (i.e. support for running SAT checks on class files generated with Java 21).

File: sat-plugin/src/main/java/org/openhab/tools/analysis/tools/PmdChecker.java [hunk 1]
```diff
@@ -69,7 +69,7 @@ public class PmdChecker extends AbstractChecker {
     @Parameter
     private List<Dependency> pmdPlugins = new ArrayList<>();
 
-    private static final String PMD_VERSION = "6.53.0";
+    private static final String PMD_VERSION = "7.0.0-rc4";
```
Comment (kaikreuzer) [inline sat-plugin/src/main/java/org/openhab/tools/analysis/tools/PmdChecker.java:72]: `7.0.0` has just been released on March 22 - you might want to update to this?

**Comment by holgerfriedrich** [comment]
@kaikreuzer When I checked a few days ago, it was still -rc4 :-)
I will quickly check if there is a functional difference and pull the final version in.

**Comment by holgerfriedrich** [comment]
A lot of changes between rc4 and final, I had to adapt a few files.
Though, now the PR is reworked and seems to work fine with the final release.

@wborn do you want to give it a try as well?

**Review by wborn — APPROVED** [review]
It still works well for me with these changes! :+1:

**Review by kaikreuzer — APPROVED** [review]
Thank you both!

**Comment by holgerfriedrich** [comment]
@kaikreuzer thanks, could you please trigger the release of 0.16?
Not sure how this works and how this gets published on the typical mvn repos....

**Comment by kaikreuzer** [comment]
@holgerfriedrich Will do. Is there any special reason why the build asks for Java 11? Or is it safe to change the constraint and build with with Java 17 instead?

**Comment by holgerfriedrich** [comment]
> @holgerfriedrich Will do. Is there any special reason why the build asks for Java 11? Or is it safe to change the constraint and build with with Java 17 instead (which seems to work well)?

I typically build with newer versions on my local machine. However, the last time we tried to upgrade to 17, we have broken the release build, see #451. Something releated to the unleash plugin.

**Comment by kaikreuzer** [comment]
@holgerfriedrich Many thanks for the reminder!
I've built it with Java11 again and here we are: https://github.com/openhab/static-code-analysis/releases/tag/0.16.0

**Comment by holgerfriedrich** [comment]
> @holgerfriedrich Many thanks for the reminder! I've built it with Java11 again and here we are: https://github.com/openhab/static-code-analysis/releases/tag/0.16.0

Thanks a lot, @kaikreuzer 🚀

**Pull request merged**